crates/ibc/src/core/ics24_host/identifier.rs

```diff
@@ -123,46 +100,95 @@ impl ChainId {
     /// assert_eq!(ChainId::is_epoch_format("chainA"), false);
     /// assert_eq!(ChainId::is_epoch_format("chainA-1"), true);
     /// assert_eq!(ChainId::is_epoch_format("c-1"), true);
+    /// assert_eq!(ChainId::is_epoch_format("-1"), false);
     /// ```
     pub fn is_epoch_format(chain_id: &str) -> bool {
-        let re = safe_regex::regex!(br".*[^-]-[1-9][0-9]*");
-        re.is_match(chain_id.as_bytes())
+        Self::split_chain_id(chain_id).1.is_some()
     }
 
     /// with_version() checks if a chain_id is in the format required for parsing epochs, and if so
     /// replaces it's version with the specified version
     /// ```
     /// use ibc::core::ics24_host::identifier::ChainId;
-    /// assert_eq!(ChainId::new("chainA".to_string(), 1).with_version(2), ChainId::new("chainA".to_string(), 2));
+    /// assert_eq!(ChainId::new("chainA", 1).with_version(2), ChainId::new("chainA", 2));
     /// assert_eq!("chain1".parse::<ChainId>().unwrap().with_version(2), "chain1".parse::<ChainId>().unwrap());
     /// ```
     pub fn with_version(mut self, version: u64) -> Self {
-        if Self::is_epoch_format(&self.id) {
-            self.id = {
-                let mut split: Vec<&str> = self.id.split('-').collect();
-                let version = version.to_string();
-                if let Some(last_elem) = split.last_mut() {
-                    *last_elem = &version;
-                }
-                split.join("-")
-            };
-            self.version = version;
+        if self.version != 0 {
```

Comment:
This is a slightly different check than previously. The previous implementation effectively checks if there's a `-` in the id, whereas this checks if version is `0` (which can occur with e.g. chain id `"mockChainNotEpochFormat"`).

I suggest we:
1. store an `is_epoch_format` boolean in the type itself, so that we only need to parse the chain ID upon construction.
  + Use this `is_epoch_format` instead of `self.version != 0`
2. Change this signature to `with_version(self, version: u64) -> Result<Self, IdentifierError>`
  + we needlessly require the caller to have a mutable reference to the object
  + we should fail if the identifier is not in epoch format, and let the caller decide what to do then. Note that this function is not used anywhere, so this change of signature should be straightforward

Reply:
`self.version == 0` implies that the id is not in epoch format so
checking it is just an optimisation¹.  As far as I can tell the old
and new code behave the same way (module panic if version doesn’t fit
`u64`).  Neither the old nor the new code will change identifier such
as `foo-bar` for example.

I can add `is_epoch_format` if you prefer but it will increase the
size of the `ChainId` structure which I feel like it could end up
hurting rather than helping.

Returning `Result` certainly makes sense though note that changing
`mut self` to `self` doesn’t actually matter for the caller.  `self`
is moved into the method.  What possibly would make most sense is
`with_version(self, u64) -> Result<Self, Self>` (or changing the
semantics such that it adds version if not already in epoch format).

Reply:
> `self.version == 0` implies that the id is not in epoch format so
checking it is just an optimisation¹.

Ah right, I didn't realize `"foo-0"` is not considered epoch format. So basically we had it wrong before, and this is correct.

> I can add is_epoch_format if you prefer

Nope, this is no longer necessary (see previous comment).

> What possibly would make most sense is `with_version(self, u64) -> Result<Self, Self>` (or changing the semantics such that it adds version if not already in epoch format).

The reason I still prefer my suggestion is that when not in epoch format, the concept of `version` in a chain ID doesn't exist. In practice, all chains now use the epoch format. I do not see a use case to "convert your chain ID to epoch format"; most likely, if someone called `with_version()` on a chain ID that is not in epoch format, it's a bug and they should fix it, as opposed to intended behavior.

Note that this whole "encode the version number in the chain id" evolved as a hack; it wasn't in the original IBC standard, and is pretty messy. Non-epoch format chain IDs are pretty much non-existent.